contracts/src/LightClient.sol

```diff
@@ -0,0 +1,157 @@
+    /// before any newer state can be accepted since the stake table commitments of that block
+    /// become the snapshots used for vote verifications later on.
+    function newFinalizedState(
+        LightClientState calldata newState,
+        IPlonkVerifier.PlonkProof calldata proof
+    ) external {
+        if (
+            newState.viewNum <= finalizedState.viewNum
+                || newState.blockHeight <= finalizedState.blockHeight
+        ) {
+            revert OutdatedState();
+        }
+        uint64 epochEndingBlockHeight = (currentEpoch + 1) * BLOCKS_PER_EPOCH - 1;
+        if (
+            finalizedState.blockHeight != epochEndingBlockHeight
+                && newState.blockHeight > epochEndingBlockHeight
+        ) {
+            revert MissingLastBlockForCurrentEpoch(epochEndingBlockHeight);
+        }
+        // format validity check
+        BN254.validateScalarField(newState.blockComm);
+        BN254.validateScalarField(newState.feeLedgerComm);
+        BN254.validateScalarField(newState.stakeTableBlsKeyComm);
+        BN254.validateScalarField(newState.stakeTableSchnorrKeyComm);
+        BN254.validateScalarField(newState.stakeTableAmountComm);
+        // sanity check on the threshold
+        if (newState.threshold <= stakeTable.totalVotingStake() * 2 / 3) {
+            revert InvalidArgs();
+        }
+
+        // check plonk proof
+        // TODO: (alex) replace the vk with the correct one
+        IPlonkVerifier.VerifyingKey memory vk = VkTest.getVk();
+        uint256[] memory publicInput = preparePublicInput(newState);
+        PlonkVerifier.verify(vk, publicInput, proof, bytes(""));
+
+        // upon successful verification, update state.
+        // If the newState is in a new epoch, only then should we increment the `currentEpoch`, and
+        // update the stake table. The `finalizedState` (before update) should have the
+        // `epochEndingBlockHeight`
+        if (finalizedState.blockHeight == epochEndingBlockHeight) {
+            bytes32 newStakeTableComm = keccak256(
+                abi.encodePacked(
+                    finalizedState.stakeTableBlsKeyComm,
+                    finalizedState.stakeTableSchnorrKeyComm,
+                    finalizedState.stakeTableAmountComm
+                )
+            );
+
+            stakeTable.advanceEpoch(newStakeTableComm);
+            currentEpoch += 1;
+        }
+
+        finalizedState = newState;
+        emit NewState(newState.viewNum, newState.blockHeight, newState.blockComm);
+    }
+
+    // === Pure or View-only APIs ===
+    /// @dev Transform a state into an array of field elements, prepared as public inputs of the
+    /// plonk proof verification
+    function preparePublicInput(LightClientState calldata state)
+        public
+        pure
+        returns (uint256[] memory)
+    {
+        uint256[] memory publicInput = new uint256[](8);
```

Comment:
Did we consider hashing these all together and unhashing in the circuit, so we just have a single public input? I don't know which would be cheaper, but I know Polygon zkEVM hashes everything before sending it into the SNARK

Reply:
Ok. Sounds good to me, let's keep it as is